sql/src/main/java/io/crate/execution/engine/aggregation/impl/CollectSetAggregation.java

```diff
@@ -100,19 +114,158 @@ public DataType partialType() {
     }
 
     @Override
-    public Set<Object> reduce(RamAccountingContext ramAccountingContext, Set<Object> state1, Set<Object> state2) {
-        for (Object newValue : state2) {
-            if (state1.add(newValue)) {
+    public Map<Object, Object> reduce(RamAccountingContext ramAccountingContext,
+                                      Map<Object, Object> state1,
+                                      Map<Object, Object> state2) {
+        for (Object newValue : state2.keySet()) {
+            if (state1.put(newValue, PRESENT) == null) {
                 ramAccountingContext.addBytes(
-                    RamAccountingContext.roundUp(innerTypeEstimator.estimateSize(newValue) + 36L) // value size + 32 bytes for entry + 4 bytes for increased capacity
+                    // value size + 32 bytes for entry + 4 bytes for increased capacity
+                    RamAccountingContext.roundUp(innerTypeEstimator.estimateSize(newValue) + 36L)
                 );
             }
         }
         return state1;
     }
 
     @Override
-    public Object[] terminatePartial(RamAccountingContext ramAccountingContext, Set<Object> state) {
-        return state.toArray(new Object[0]);
+    public Object[] terminatePartial(RamAccountingContext ramAccountingContext, Map<Object, Object> state) {
+        return state.keySet().toArray(new Object[0]);
+    }
+
+    @Override
+    public boolean isRemovableCumulative() {
+        return false;
+    }
+
+    /**
+     * collect_set implementation that is removable cumulative. It tracks the number of occurrences for every key it
+     * sees in order to be able to only remove a value from the aggregated state when it's occurrence count is 1.
+     *
+     * eg. for a window definition of CURRENT ROW -> UNBOUNDED FOLLOWING for {1, 2, 2, 2, 3 } the window frames and
+     * corresponding collect_set outputs are:
+     *  {1, 2, 2, 2, 3}  - [1, 2, 3]
+     *  {2, 2, 2, 3}     - [2, 3]
+     *  {2, 2, 3}        - [2, 3]
+     *  {2, 3}           - [2, 3]
+     *  {3}              - [3]
+     */
+    private static class RemovableCumulativeCollectSet extends AggregationFunction<Map<Object, Object>, Object[]> {
+
+        private static final transient int[] ONE_OCCURENCE = new int[]{1};
+
+        private final SizeEstimator<Object> innerTypeEstimator;
+
+        private final FunctionInfo info;
+        private final DataType partialReturnType;
```

Comment:
```suggestion
        private final DataType partialType;
```

The partial type is not restricted to return values, but also used as inputs to reduce. So I think `returnType` is miss-leading.